ksqldb-parser/src/main/java/io/confluent/ksql/parser/properties/with/CreateSourceProperties.java

```diff
@@ -127,71 +123,50 @@ public String getKafkaTopic() {
   }
 
   public Optional<Integer> getKeySchemaId() {
-    return Optional.ofNullable(props.getInt(CommonCreateConfigs.KEY_SCHEMA_ID));
+    return SourcePropertiesUtil.getKeySchemaId(props);
   }
 
   public Optional<Integer> getValueSchemaId() {
-    return Optional.ofNullable(props.getInt(CommonCreateConfigs.VALUE_SCHEMA_ID));
+    return SourcePropertiesUtil.getValueSchemaId(props);
   }
 
-  public Optional<FormatInfo> getKeyFormat(final SourceName name) {
+  public Optional<FormatInfo> getKeyFormat(final SourceName name, final KsqlConfig config) {
```

Comment:
That's a valid option, yeah. I didn't do it because I wanted the change to be transparent to users -- if we add a property into CreateSourceProperties and CreateSourceAsProperties to represent this config, then it will be serialized as part of the statement string (visible when describing sources and queries) and might lead to confusion for users. On the other hand, not needing to pipe an extra config into the analyzer is definitely cleaner code-wise. 

It might be an option to not expose this config as a property in CreateSourceProperties and CreateSourceAsProperties and instead introduce specialized methods in CreateSourceProperties and CreateSourceAsProperties to set the format property for the protobuf format (with an injector, as you suggested). Doing so means we wouldn't need to expose the property to users as part of the statement string, but would be a bit of a code smell in CreateSourceProperties and CreateSourceAsProperties. I can look into this if you think this is cleaner than passing the config into the analyzer.

Reply:
I see. If we need to create a new property in `CreateSourceProperties` and `CreateSourceAsProperties`, it feels like a ksqlDB internal property not to be used by the user (or it should be used by user replacing this config `ksql.protobuf.unwrap.primitives`?). Is there a mechanism to define internal property? If not,  to avoid displaying the property to user, can we have a blocklist of properties which won't be serialized to statement string? 

Another option: I feel passing `KsqlConfig` to `CreateSourceProperties` to get some property is similar to have another set of properties in `CreateSourceProperties` which is not set from `CT/CS` command. Maybe you can store `KsqlConfig` in `CreateSourceProperties` in the beginning then? Or create another set of properties from `KsqlConfig` and store it in `CreateSourceProperties`?

Just to make code cleaner and have less changes in this PR. WDYT?

Reply:
Hi @lihaosky , I started drafting changes to remove the new KsqlConfig property in favor of storing information in CreateSourceProperties and CreateSourceAsProperties instead. The approach is to add an injector that updates CreateSourceProperties and CreateSourceAsProperties for new queries only to indicate that any protobuf sources should use the new converter behavior (unwrap wrapped primitives). This then gets written into the command topic as the new protobuf format property. You can see the changes I drafted here: https://github.com/vcrfxia/ksql/tree/protobuf-primitives-rebase-new

I think the new approach is OK (cleaner than piping a new config into the analyzer) but the trouble is that the QTT framework is not set up to play nicely with this. There are multiple places in the QTT execution code where statement injectors are manually applied to massage statements, e.g., [here](https://github.com/confluentinc/ksql/blob/177fecc16a177113ce45120470ecc307dbfd282e/ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestCaseBuilderUtil.java#L229). In these places, we need to be able to decide whether the new injector should be applied or not. We can't easily do this without adding a new config since it doesn't make sense to decide whether to apply the new injector or not based on whether a format property exists in the plan (there's no piping set up for this, and I don't think we'd want to do it anyway).

I'm disappointed with this conclusion but I think it means that even with this new approach, we'd still have to add a new KsqlConfig (in order to get QTT to play nicely). I don't think it makes sense to add both a new config and a new injector -- we should just use one to be consistent.

What do you think? Should we stick with the old approach (new config and no injector)? Have I missed something?

Reply:
Discussed offline. We will keep the `KsqlConfig` and `Injector` and populate the property in `Injector` based on `KsqlConfig`. Hopefully this can resolve the QTT problem.

BTW, injectors are also called here in QTT: https://github.com/confluentinc/ksql/blob/177fecc16a177113ce45120470ecc307dbfd282e/ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestExecutorUtil.java#L487-L502

Reply:
I've made enough changes to this implementation that I think it makes sense to open a new PR instead -- see https://github.com/confluentinc/ksql/pull/8934.